### ISSUE TITLE
Can you please elaborate how to call this function ?

### ISSUE THREAD

**Issue by lingyuj** [zhangyuang/node-ffi-rs#26]
I can't figure out how to call the below function, since there are not many examples. Can you please show me how to call it with ffi-rs? Thanks in advance.

typedef struct
{
        uint8_t bytes[16];
} my_struct_t;

extern "C" int func1 (
        const my_struct_t* param1,
        const char* certs[],
        uint32_t ncerts,
        uint32_t* cert_index
);

**Comment by zhangyuang** [comment]
At first, you should figure out each function params type.

In this case

The first param is a pointer to a structure. second param is string array, the third param is u32 number, the fourth param is a pointer to u32.

There are two ways to create struct pointer both `createPointer` or pass object directly to `ffi-rs`.

Here is an example 

```
const cert_index = 100
  const structPointer = createPointer({
    paramsType: [{
      bytes: DataType.U8Array
    }],
    paramsValue: [[]]
  })
  const intPointer = createPointer({
    paramsType: [DataType.I32],
    paramsValue: [cert_index]
  })
  load({
    library: '',
    funcName: 'func1',
    paramsType: [
      DataType.External,
      DataType.StringArray,
      DataType.I32,
      DataType.External
    ],
    paramsValue: [
      unwrapPointer(structPointer),
      ["foo"],
      100,
      intPointer[0]
    ],
    retType: DataType.I32
  })
```

**Comment by lingyuj** [comment]
Here is my code. The c function can be called now but threw an exception: thread '<unnamed>' panicked at 'unsupported params type Undefined', src/utils/dataprocess.rs:168:14

  const cpu_svn_value = {
    bytes: Buffer.from(cpu_svn, 'hex')
  }
  const cpusvnPointer = createPointer({
    paramsType: [{
      bytes: DataType.U8Array
    }],
    paramsValue: [cpu_svn_value]
  })
  console.log(cpu_svn_type)      // output is:  { byteArray: { type: 10, length: 16, ffiTypeTag: 'array' } }
  console.log(cpu_svn_value)     // output is : { bytes: <Buffer 05 05 13 15 03 ff 00 ff 00 00 00 00 00 00 00 00> }
  console.log(cpusvnPointer)    // output is : [ [External: 78cd9c0] ]

  const cert_index = 0
  const best_index_ptr = createPointer({
    paramsType: [DataType.I32],
    paramsValue: [cert_index]
  })

  const ret = load({
    library: 'libPCKCertSelection', // path to the dynamic library file
    funcName: 'pck_cert_select', // the name of the function to call
    retType: DataType.I32, // the return value type
    paramsType: [DataType.External, DataType.I16, DataType.I16, DataType.String, DataType.StringArray, DataType.I32, DataType.External], // the parameter types
    paramsValue: [cpusvnPointer[0], my_pce_svn, my_pce_id, tcb_info, pem_certs, ncerts, best_index_ptr[0]] // the actual parameter values
  })

**Comment by zhangyuang** [comment]
DataType has not support `i16`, you can get type hints by writing typescript

**Comment by lingyuj** [comment]
> DataType has not support `i16`, you can get type hints by writing typescript

ok, thanks. I will try again.

**Comment by lingyuj** [comment]
> BTW, may be you need use `unwrapPointer` on `cpusvnPointer[0]` for get the inner struct pointer

It seems no unwrapPointer is needed. The code worked after I changed I16 to I32. But how can I get the value out of best_index_ptr after the call?

**Comment by zhangyuang** [comment]
Use `restorePointer` to get the data pointed to by the pointer

**Comment by lingyuj** [comment]
> BTW, may be you need use `unwrapPointer` on `cpusvnPointer[0]` for get the inner struct pointer

You are correct that the cpusvn is not passed to the c library correctly. I tried different methods with no success.
 pass the struct directly:
```
  const cpu_svn_value = {
    bytes: Buffer.from(cpu_svn, 'hex')
  }
  const ret = load({
    library: 'libPCKCertSelection', // path to the dynamic library file
    funcName: 'pck_cert_select', // the name of the function to call
    retType: DataType.I32, // the return value type
    paramsType: [{bytes: DataType.U8Array}, DataType.I32, DataType.I32, DataType.String, DataType.StringArray, DataType.I32, DataType.External], // the parameter types
    paramsValue: [cpu_svn_value, my_pce_svn, my_pce_id, tcb_info, pem_certs, ncerts, best_index_ptr[0]] // the actual parameter values
  })
```
or use a pointer:
```
  const cpu_svn_value = {
    bytes: Buffer.from(cpu_svn, 'hex')
  }
  const cpusvnPointer = createPointer({
    paramsType: [{
      bytes: DataType.U8Array
    }],
    paramsValue: [cpu_svn_value]
  })
  const ret = load({
    library: 'libPCKCertSelection', // path to the dynamic library file
    funcName: 'pck_cert_select', // the name of the function to call
    retType: DataType.I32, // the return value type
    paramsType: [DataType.External, DataType.I32, DataType.I32, DataType.String, DataType.StringArray, DataType.I32, DataType.External], // the parameter types
    paramsValue: [cpusvnPointer[0], my_pce_svn, my_pce_id, tcb_info, pem_certs, ncerts, best_index_ptr[0]] // the actual parameter values
  })
```
For the second approach, if I use unwrapPointer(cpusvnPointer[0]), the error log says "Error: Given napi value is not an array"

Here is my c function definition:
```
typedef struct
{
	uint8_t bytes[16];
} cpu_svn_t;

pck_cert_selection_res_t pck_cert_select (
        const cpu_svn_t* platform_svn,
        uint16_t pce_isvsvn,
        uint16_t pce_id,
        const char* tcb_info,
        const char* pem_certs[],
        uint32_t ncerts,
        uint32_t* best_cert_index )
```

**Comment by zhangyuang** [comment]
I will handle to problem at later

**Comment by zhangyuang** [comment]
The problem has been fixed in `ffi-rs@1.0.59`.

You need to define platform_svn field in paramsType like this

```js
{

 paramsType: [{
      bytes: arrayConstructor({
        type: DataType.U8Array,
        length: 16,
        dynamicArray: false
      }),
    }],

}
```

The behavior of dynamicArray field can refer the latest update of [readme](https://github.com/zhangyuang/node-ffi-rs?tab=readme-ov-file#struct)

**Comment by lingyuj** [comment]
> The problem has been fixed in `ffi-rs@1.0.59`.
> 
> You need to define platform_svn field in paramsType like this
> 
> ```js
> {
> 
>  paramsType: [{
>       bytes: arrayConstructor({
>         type: DataType.U8Array,
>         length: 16,
>         dynamicArray: false
>       }),
>     }],
> 
> }
> ```
> 
> The behavior of dynamicArray field can refer the latest update of [readme](https://github.com/zhangyuang/node-ffi-rs?tab=readme-ov-file#struct)


> The problem has been fixed in `ffi-rs@1.0.59`.
> 
> You need to define platform_svn field in paramsType like this
> 
> ```js
> {
> 
>  paramsType: [{
>       bytes: arrayConstructor({
>         type: DataType.U8Array,
>         length: 16,
>         dynamicArray: false
>       }),
>     }],
> 
> }
> ```
> 
> The behavior of dynamicArray field can refer the latest update of [readme](https://github.com/zhangyuang/node-ffi-rs?tab=readme-ov-file#struct)

Awesome! Thanks for the quick response. Will test it ASAP.

**Comment by lingyuj** [comment]
Confirmed the issue has been fixed in ffi-rs@1.0.59. This project is awesome and thank you for the hard work.
